### DISCUSSION TITLE
Idle power consumption

### DISCUSSION THREAD

**Discussion by luar123** [CarlosDerSeher/snapclient#43] [Ideas]
Hey everyone, I have been looking into reducing idle current and want to share findings and get some new ideas.
My setup is not optimized and consists of a ESP32 DevKit with pcm5102a dac. I will always compare current floor. There are higher peaks when e.g. wifi is transmitting, so to be more accurate we would need to integrate over some time. But for a first comparison floor is good enough guess.

Default idle current is around 83mA, not too bad but still, the device is doing almost nothing. 
Power management needs to be enabled through menuconfig (components->powermanagement) and then activated in the code. With this CPU frequency scaling can be activated. Automatic light sleep mode requires additionally tickless idle freeRTOS mode (components->FreeRTOS->tickless idle).

Using this, idle current can be decreased to:
- CPU freq scaling only: 59mA
- additional light sleep: 22mA 

So I configured frequency scaling and if available automatic sleep mode.
Then I used initially a lock to deactivate it when music is playing (I set the lock in the `audio_set_mute` function). This is working after boot-up, but once the device is muted and un-muted again, it is out of sync. I figured that some timers got confused by the frequency scaling and tried to automatically stop the player task whenever the device is muted or music is stopped and then re-initialize it when music resumes. I got this working after some try and error and with light sleep and freq scaling it seems to keep the sync. However, for freq scaling only it gets out of sync again. Additionally I am not sure if stopping the player task is the right thing to do. It should be enough to reset the sync when music resumes.

You can find a proof-of-concept in https://github.com/luar123/snapclient/tree/power_reduction
(You need to configure your board and enable power management through menuconfig.)

I have two questions:

- Any thoughts on how to best reset the sync after power management is disabled / music is resumed?
- According to espressif docu: `Enabling power management features comes at the cost of increased interrupt latency`: 0.2-40uS. Do you think (or can someone test and measure) this will have a significant influence on the overall sync?

**Comment by CarlosDerSeher** [comment]
Great you are getting this started, this has been on my task list for a long time just not a real high priority. I couldn't make it work and keep up sync right away so pushed it aside.

One thing that could be an issue in player task is the hardware timer used for inital sync. I believe its clock is influenced by frequency scaling?! Maybe take a lock while waiting for initial sync helps?

The other thing which could be problematic is a jittery latency measurement because of power management. Have a look at esp_timer and how it is affected by PM. This is used for the time ping-pong with the server. I think relevant lwip send functions are called by periodic timer callbacks, maybe a bunch are missed because of sleep and then sent in a row. 

I don't think 40us will be a problem in a multiroom setup. For stereo speakers they probably are. If you have a oscilloscope you could use two boards and and measure the same channel on both to see sync. Use a metronome signal 120bpm or similar.

**Reply by luar123** [reply]
Yes, esp_timer is definitely influenced by the frequency scaling and I believe this messes with the sync in the normal implementation. With the modified player task a lock is taken whenever the player task is running and the timer gets initialized new each time.

Light sleep shouldn't be a problem as it is supposed to wake up when a timer callback is due, but I need to look further into the latency measurement. Where exactly is this ping-pong happening?

I will try to do some measurements in the next days.

**Reply by CarlosDerSeher** [reply]
Ot is somewhere in main.c, look for
`SNAPCAST_MESSAGE_TIME`

And when you say it gets reinitialized, do you mean it's counting value is reset? If so then there is your problem

**Reply by CarlosDerSeher** [reply]
Maybe it will be agood idea to also take a apb and/or light sleep lock during the time while waiting for the time message response from the server

**Reply by CarlosDerSeher** [reply]
I did set APB max to 80Mhz once and it did work flawless, so maybe disabling DFS completely and just setting the clock to its lowest setting will be a good solution?

**Comment by luar123** [comment]
Finally, I was able to get some useful measurements. My reference is a x86 server that generated a lot of noise on the scope, so I needed some additional filtering. I really need to get a second esp+dac

Anyway, here are the results, energy average over 0.5-2h:
1. master branch, without PM:
  0.84W when playing
  0.4W when mute
  time difference +/-0.2ms
  after un-mute the diff is higher, and goes down to the usal range within a few seconds
2. my proof of concept branch, without pm, cpu freq=240MHz:
  same as master
3. my proof of concept branch, without pm, cpu freq=80MHz:
  0.78W when playing
  0.34W when mute
  same sync/behaviour as master
4. my proof of concept branch, with pm DFS+Lsleep, cpu freq=240MHz:
  0.84W when playing (pm disabled)
  0.2W when mute
  time diff +/-0.2ms
  after un-mute the diff is much higher (sometimes >>10ms), and takes a few minutes to go down to the usual range

Conclusions so far:

- your syncing is really good :-)
- CPU scaling gives no big power saving
- light sleep has biggest impact
- disable PM while playing
- when PM is enabled sync is not kept, solutions:
  - try to keep sync,maybe by taking a lock while getting time from server
  - force to reset/recover the sync once pm is disabled

Not sure what's the best solution. When I have time I will experiment more.

**Reply by CarlosDerSeher** [reply]
Reads like you are putting lots of effort in there, thumbs up for that 👍 and yes I also put in a lot of effort to sync so thanks :)

So here it goes:
Another thing you should consider is the  server not sending audio. Clients should enter power management in that case too. This is the situation for my setup at least. Maybe just kill player task.

A possible solution for the time sync problem could be to clear the latency buffer and do a new initial sync with the server. This will put a bit more stress on the network but could be a fast solution for now.

**Reply by CarlosDerSeher** [reply]
Found another interesting thing

https://github.com/bitluni/ULPSoundESP32

Uses ULP for audio out. Not sure if this could be done for i2s audio too.

**Reply by luar123** [reply]
Not sure if ULP could theoretical work, but I don't want to look into it. Ok, maybe on ESP32-S2, but on the others you have to code in assembly...

**Reply by CarlosDerSeher** [reply]
I don't really think ULP would be practical or even possible. APLL will always need to run and I2S / DMA, from what I read, isn't able to wake up the processor from light sleep.

**Comment by luar123** [comment]
Got it working now.

> So here it goes: Another thing you should consider is the server not sending audio. Clients should enter power management in that case too. This is the situation for my setup at least. Maybe just kill player task.

That's actually my current implementation. Muting is ignored, but when the pcm queue is empty, the player_task will kill itself.

> A possible solution for the time sync problem could be to clear the latency buffer and do a new initial sync with the server. This will put a bit more stress on the network but could be a fast solution for now.

Thanks, that was the missing part :)
Also had to set `timeout = FAST_SYNC_LATENCY_BUF` in order to keep the delay until playing starts low.

So my current implementation is:

- init player_task (but don't start)
- when pcm chunks are inserted player_task is started
  - take apb lock
  - start timer
  - reset latency buffer
  - create settings/pcm queue
- when pcm queue is empty, stop player_task
  - stop timer
  - delete queues
  - release lock
  - delete task

This is working fine, but I don't like the implementation and won't create a PR. Now I would keep player_task running and just take/release locks, resync, block task etc when needed.
Additionally, I think it would be a good idea to put all audio_HAL handling (mute, volume) in one task. Player_task already handles all i2s and also knows when there is nothing to play, so this would be a good fit. Finally, player task could inform http_task of its state (playing/paused) in order to get FAST_SYNC_LATENCY_BUF as fast as possible.

If you agree, I would implement this and create a PR for it. Might take a while, though...my time is limited :D

**Reply by CarlosDerSeher** [reply]
> when pcm queue is empty, stop player_task 

Maybe increase the timeout of queue to 5s as badaix has it in original snapcast

>  Also had to set timeout = FAST_SYNC_LATENCY_BUF in order to keep the delay until playing starts low.

Did you also look into taking a light sleep lock during time sync ping-pong so sync buffer maybe wouldn't need to be reset? It could put quite some stress on the server if >=10 clients are flooding it with time sync messages at start of playback.

Apart from that your suggestions sound good.

**Reply by luar123** [reply]
That is a good point, have not tried yet.

**Reply by CarlosDerSeher** [reply]
Also I recommend using non IDF specific API where you can, so code can be ported to other platforms more easily. So probably a freertos queue from player to http task will be the way to go and not a an event loop for the muting and volume part.

**Reply by luar123** [reply]
>Did you also look into taking a light sleep lock during time sync ping-pong so sync buffer maybe wouldn't need to be reset?

Unfortunately this did not help at all.
It seems the esp time is always drifting a bit, between 0 and 100µs per second in normal operation, but with pm enabled its more between 2ms and 30ms per second (depending on the settings). This seems to be related to the 150khz/8.5MHz oscillator in light sleep and due to DFS.
As a result the latency value is changing fast. This persists after music is resumed until half of the median buffer is filled with new values. Depending on the actual settings this could lead to hard resyncs or in the worst case an age slightly smaller than 10ms that is slowly recovered by apll tuning.

To get the details, more investigations are necessary, but for now the easiest solution is to reset the latency buffer.  Instead of using `FAST_SYNC_LATENCY_BUF` we could also reduce the size of the median filter and/or start playing before the buffer is completely full. For my network, even very small filters with only 9 elements were ok.

As an alternative we could sync time each time music is resumed, e.g. with ntp. Also an external 32kHz crystal could improve the drift during sleep.

**Reply by CarlosDerSeher** [reply]
> or start playing before the buffer is completely full

I'd opt for that solution. So do a fast sync for lets say 9 entries and then change to 1s slow sync interval and fill the buffer.

> Also an external 32kHz crystal could improve the drift during sleep.

I have some boards which got 32khz crystal attached. I'll test your branch when I got some spare time to see how this goes.

> As an alternative we could sync time each time music is resumed, e.g. with ntp.

Also I'd rather not make NTP a requirement as this would make an active internet connection a must have.

> but with pm enabled its more between 2ms and 30ms per second (depending on the settings).

Just wondering, how did you get those numbers?

**Reply by luar123** [reply]
ok, I agree.
> I have some boards which got 32khz crystal attached. I'll test your branch when I got some spare time to see how this goes.

That would be interesting. I got the most stable clock when setting freq_min=freq_max=80MHz in main.c and also cpu freq=80MHz in menuconfig. Basically disabling DFS completely.
To monitor the time drift, enable logging of the latency in `SNAPCAST_MESSAGE_TIME:` (more noisy) or in `player_latency_insert()` (short delay with a small median filter). Then you get one log per second and if pm is enabled for me the value was decreasing fast in the range I wrote above.

**Reply by CarlosDerSeher** [reply]
So I tried building and flashing your branch from my ubuntu 23.10 machine but I am running into following error during flashing through `idf.py -b 1000000 flash`

`esptool write_flash: error: argument <address> <filename>: Detected overlap at address: 0x8000 for file: partition_table/partition-table.bin`

It builds OK but complains that bootloader overlaps with partition table. I checked sdkconfig and compared it to my own default one and it seems there hasn't been any change that could cause this. The only way I could make it flash successfully is to change default log level in bootloader from INFO to WARNING. Did you see this too?

I've been struggling with this, so can't really report on anything else yet...

**Reply by CarlosDerSeher** [reply]
Ok, so it seems to bee related to RTC 32.768kHz clock source settings, got it running now

> I (24960) PLAYER: PM disabled
I (24971) PLAYER: player_setup_i2s: dma_buf_len is 576, dma_buf_count is 4
I (24972) c_I2S: DMA Malloc info, datalen=blocksize=2304, dma_buf_count=4
I (24979) c_I2S: APLL: Req RATE: 44100, real rate: 44099.988, BITS: 16, CLKM: 1, BCK_M: 8, MCLK: 11289597.000, SCLK: 1411199.625000, diva: 1, divb: 0
I (24989) GENERIC_BOARD: I2S0, MCLK output by GPIO3
E (24995) PLAYER: unable to register timer 1 callback
I (25000) PLAYER: Start player_task
I (25005) PLAYER: started sync task
I (25009) PLAYER: start player done
I (25014) c_I2S: APLL: Req RATE: 44100, real rate: 44099.988, BITS: 16, CLKM: 1, BCK_M: 8, MCLK: 11289597.000, SCLK: 1411199.625000, diva: 1, divb: 0
I (25027) PLAYER: created new queue with 28
I (25452) SC: latency buffer not full
I (25639) SC: latency buffer full
I (26354) PLAYER: initial sync age: 20us, chunk duration: 26122us

Note the error above

> E (24995) PLAYER: unable to register timer 1 callback

Also after starting playback I get continuous resyncs and no playback

> W (59119) PLAYER: RESYNCING HARD 1: age 8667us, latency 862839681851us, free 176424, largest block 110592, rssi: -46
W (59152) PLAYER: RESYNCING HARD 1: age 15391us, latency 862839681851us, free 176428, largest block 110592, rssi: -46
W (59186) PLAYER: RESYNCING HARD 1: age 22970us, latency 862839681851us, free 174952, largest block 110592, rssi: -42
W (59217) PLAYER: RESYNCING HARD 1: age 27419us, latency 862839681851us, free 179712, largest block 110592, rssi: -42
W (59251) PLAYER: RESYNCING HARD 1: age 35764us, latency 862839681851us, free 176552, largest block 110592, rssi: -42
W (59286) PLAYER: RESYNCING HARD 1: age 44309us, latency 862839681851us, free 178024, largest block 110592, rssi: -44
W (59319) PLAYER: RESYNCING HARD 1: age 51490us, latency 862839681851us, free 176540, largest block 110592, rssi: -44
W (59353) PLAYER: RESYNCING HARD 1: age 59279us, latency 862839681779us, free 176572, largest block 110592, rssi: -44
W (59388) PLAYER: RESYNCING HARD 1: age 67727us, latency 862839681779us, free 176560, largest block 110592, rssi: -44
W (59421) PLAYER: RESYNCING HARD 1: age 74784us, latency 862839681779us, free 174912, largest block 110592, rssi: -44

I don' have time to look into it further right now though

**Reply by CarlosDerSeher** [reply]
OK, seems it is also drifting with 32kHz attached. So no difference after a quick test.

Not sure why these resyncs where happening before

**Reply by luar123** [reply]
Thanks for testing. I never tried the 32kHz setting, as I don't have hardware.
But I could try to set it and flash it, just to see if I get the same error.

What PM/Frequency setting did you use for that test?

`E (24995) PLAYER: unable to register timer 1 callback` this error is normal. It is not causing any issues and I was just too lazy to fix it.
The resyncs are strange, but is it working now?

**Reply by CarlosDerSeher** [reply]
> But I could try to set it and flash it, just to see if I get the same error.

Never mind it will fall back to 150khz oscillator if 32khz isn't actually present. There are some settings which are related to drive current of 32khz and enabling these seem to make bootloader take up more space than 32k. Could be resolved by adapting partitions.csv and moving all up  by 0x1000 bytes. Didn't need them so all is OK.

> ... but is it working now?

Yes it works, but it's like you said, if left alone in light sleep for a while latency measurement will fail for some reason. Not sure why, not sure if I'll look into it. Probably though as I still think this can be made working.

> What PM/Frequency setting did you use for that test?

I set f_max=240MHz and f_min = 40MHz because it is the XTAL frequency. You had it set to 10MHz, are you sure this is even possible, [documentation](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/power_management.html) says so but I am not sure about it (see the table at `Power Management Locks`).

**Reply by luar123** [reply]
Ok, but you enabled both power management and FreeRTOS->tickless idle in menuconfig?

I set it to 10MHz because that's what is stated in the documentation and it is working. I never tracked the actual frequency though. And tbh there is no big difference in power consumption between f_min=40 and f_min=10MHz.

**Reply by CarlosDerSeher** [reply]
Yes I enabled both. Personally I think DFS is more of a hassle for what it gives... As you already noticed yourself, light sleep is the way to go to get the most power saving.

**Reply by luar123** [reply]
> Additionally, I think it would be a good idea to put all audio_HAL handling (mute, volume) in one task. Player_task already handles all i2s and also knows when there is nothing to play, so this would be a good fit.

While working on this I am not so sure anymore if it is a good idea. Depending on the DAC it could include some i2c writes. Do you think it is ok to have such a delay in the player task? Or would it be better to keep it in the http task?

**Reply by CarlosDerSeher** [reply]
I can't remember if these are blocking writes or not. If the i2c driver handles the communication in its own task and the jobs are only put to a queue from player task it is ok but every delay in player task will affect sync because at the beginning of the loop we read server_now and at the end there is the control code.

Another thing that came to my kind is DAC current consumption. Not sure if the supported codecs are all send to sleep while muted or when no chunks are received. I guess it is done for the lyrat but if you have muse you could also evaluate if there can be saved some more mA :)

**Reply by luar123** [reply]
It seems i2c is blocking. On the other hand, handling settings like muting/volume change would happen at the very beginning of the loop before server_now().

I have looked only into pcm5102a which is set to sleep mode when it is muted. So yes, could be missing for some DACs.

**Reply by CarlosDerSeher** [reply]
> On the other hand, handling settings like muting/volume change would happen at the very beginning of the loop before server_now().

Then we should get away with doing these i2c calls I think (didn't look at the code though).

**Reply by CarlosDerSeher** [reply]
On the other hand syncing relies on getting the next chunk as fast as possible after i2s_write() and evaluate if we are in time or not. So any delay in this loop is really bad. A possibility would be to put a timeout in i2s_write() and do those i2c calls on i2s write timeout and then try again to i2s_write? Not sure if this is a good solution as it could still introduce a delay from time to time as we have no way of knowing how much time is left until i2s write will succeed with requested amount of bytes.

Probably the median filter will make this delay even unnoticeable...

**Comment by DerPicknicker** [comment]
I have nothing to commit but a question in general. Is there a way to power down the dacs if they aren't used for some time..? 
My board supports powering down the dac via an single pin / i2c commands. This should be also part of the energy saving mechanism. I would guess that you can save 0,5w-1w on dacs with built-in amplifiers.

**Reply by CarlosDerSeher** [reply]
That's what this function actually is for I guess

```
esp_err_t audio_hal_ctrl_codec(
  audio_hal_handle_t audio_hal,
  audio_hal_codec_mode_t mode,
  audio_hal_ctrl_t audio_hal_state
)
```

https://github.com/CarlosDerSeher/snapclient/blob/master/components%2Faudio_hal%2Faudio_hal.c#L98

**Reply by DerPicknicker** [reply]
Okay.. but this makes for me nothing clearer. How does this function know that the dac has a dedicated power down pin or a specific i2c command to shutdown the amplifier?

**Reply by CarlosDerSeher** [reply]
Please check out audio hal from ADF2.4.

There is also a function called pa_enable or similar. audio_hal_ctrl_codec would need to be called just before entering power saving. In this function you'll have to set the gpio as needed.

For some dac's it's enough to mute them so they enter low power mode.

Disclaimer: all these things I recall from heart so don't take this for a 100% sure. Please do some research around ADF and how they do it for their boards.

**Reply by luar123** [reply]
Ok, so basically this function should be called every time `audio_hal_mute()` is called. We could use `audio_set_mute()` for that.

**Reply by CarlosDerSeher** [reply]
I think it is called `audio_hal_set_mute()` And yes I think muting would be a good way to put the DACs to sleep don't you. Also the PA could be powered down if we mute. In later ADF revisions there is also a `pa_enable` function but I'd just opt for `mute == sleep` and implement the functions this way

**Reply by DerPicknicker** [reply]
I think the main question is: 
What must the driver do, to support the power down by the function mentioned above? Is the power down in the mute function located?

The driver of the dacs must somehow support the powerdown functionalities or not?

**Reply by CarlosDerSeher** [reply]
These functions need to be called from player context, when pcm chunks start arriving and if no pcm chunks are available for x seconds
